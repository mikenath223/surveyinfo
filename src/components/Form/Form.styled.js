import styled from 'styled-components';

export const SurveyIntro = styled.h2`
  font-size: ${({theme}) => theme.fontWeights.large};
  font-weight: ${({theme}) => theme.fontWeights.semiBold};
  text-align: center;
  line-height: 2.5;
  margin-bottom: 50px
`