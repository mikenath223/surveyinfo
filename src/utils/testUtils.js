import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import { globalTheme } from 'styles/global';
import userDefTheme from 'styles/themeModes';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import store from 'store'

export const withThemeProvider = (component) =>
  render(<ThemeProvider theme={globalTheme}>
    {component}
  </ThemeProvider>)

export const withReduxThemeProvider = (component) =>
  render(<Provider store={store}>
    <ThemeProvider theme={globalTheme}>{component}</ThemeProvider>
  </Provider>);

export const withReduxProvider = (component) =>
  render(<Provider store={store}>
    {component}
  </Provider>)

export const withProviders = (component, themeMode) =>
  render(<Provider store={store}>
    <MemoryRouter>
      <ThemeProvider theme={userDefTheme[themeMode]}>
        {component}
      </ThemeProvider>
    </MemoryRouter>
  </Provider>)
