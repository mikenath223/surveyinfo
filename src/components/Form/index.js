import React from "react";
import { Formik } from "formik";
import { Form, InputGroup, Row, Col, Button } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { SurveyIntro } from './Form.styled'
import * as Yup from "yup";

const schema = Yup.object({
  message: Yup.string().required(),
  date: Yup.date().required(),
  layout: Yup.string(),
  color: Yup.string(),
  text: Yup.string()
});

export default function SurveyForm({ handleSubmit }) {
  return (
    <Formik
      validationSchema={schema}
      onSubmit={(vals) => handleSubmit(vals)}
      initialValues={{
        text: "REGULAR", color: "LIGHT", layout: "LIST",
        message: '', date: ''
      }}
    >
      {({
        handleSubmit,
        handleChange,
        handleBlur,
        values,
        touched,
        isSubmitting,
        isValid,
        errors,
      }) => (
          <Form noValidate onSubmit={handleSubmit} className="ml-lg-5 m-4">
            <SurveyIntro>Take the survey</SurveyIntro>
            <Form.Group as={Row} className="pb-4 mx-0" controlId="validationFormikMessage1">
              <Form.Label>Message</Form.Label>
              <InputGroup>
                <Form.Control
                  as="textarea"
                  name="message"
                  className="mh-100"
                  rows="7"
                  data-testid="messageInput"
                  spellCheck="true"
                  value={values.message}
                  onBlur={handleBlur}
                  onFocus={handleBlur}
                  onChange={handleChange}
                  isValid={touched.message && !errors.message}
                  isInvalid={errors.message ? touched.message : (isSubmitting && errors.message)}
                />
                <Form.Control.Feedback data-testid="messageError" tooltip type="invalid">{errors.message}</Form.Control.Feedback>
                <Form.Control.Feedback tooltip>Looks good!</Form.Control.Feedback>
              </InputGroup>
            </Form.Group>
            <Form.Group as={Row} className="pb-4 mx-0" controlId="validationFormikDate2">
              <Form.Label className="col-12 pl-0">Date</Form.Label>
              <InputGroup className="p-0 col-5">
                <Form.Control
                  type="date"
                  name="date"
                  value={values.date}
                  onChange={handleChange}
                  onBlur={handleBlur}
                  isValid={touched.date && !errors.date}
                  isInvalid={errors.date ? touched.date : (isSubmitting && errors.date)}
                />
                <Form.Control.Feedback tooltip>Looks good!</Form.Control.Feedback>
                <Form.Control.Feedback type="invalid" tooltip>{errors.date}</Form.Control.Feedback>
              </InputGroup>
            </Form.Group>

            <Form.Row className="pb-4 px-0 mx-0 mt-4">
              <Form.Label as={Col} className="col-12 pl-0">Customize your theme</Form.Label>
              <Row className="col-12 mx-0 border border-1 mt-1 p-2">
                <Form.Group as={Col} md="4">
                  <Form.Label className="font-weight-bold">Layout</Form.Label>
                  <InputGroup>
                    <div role="group" aria-labelledby="my-radio-group">
                      <Form.Check
                        type="radio"
                        id="layout"
                        name="layout"
                        label="List"
                        onChange={handleChange}
                        defaultChecked
                        value="LIST"
                      />
                      <Form.Check
                        type="radio"
                        id="layout"
                        name="layout"
                        label="Grid"
                        value="GRID"
                        onChange={handleChange}
                      />
                    </div>
                  </InputGroup>
                </Form.Group>

                <Form.Group as={Col} md="4">
                  <Form.Label className="font-weight-bold">Color Mode</Form.Label>
                  <InputGroup>
                    <div role="group" aria-labelledby="my-radio-group">
                      <Form.Check
                        type="radio"
                        id="color"
                        name="color"
                        label="Light Mode"
                        value="LIGHT"
                        onChange={handleChange}
                        defaultChecked
                      />
                      <Form.Check
                        type="radio"
                        id="color"
                        name="color"
                        label="Dark Mode"
                        value="DARK"
                        onChange={handleChange}
                      />
                    </div>
                  </InputGroup>
                </Form.Group>

                <Form.Group as={Col} md="4">
                  <Form.Label className="font-weight-bold">Text Size</Form.Label>
                  <InputGroup>
                    <div role="group" aria-labelledby="my-radio-group">
                      <Form.Check
                        type="radio"
                        id="text"
                        name="text"
                        label="Regular"
                        value="REGULAR"
                        defaultChecked
                        onChange={handleChange}
                      />
                      <Form.Check
                        type="radio"
                        id="text"
                        name="text"
                        label="Large"
                        value="LARGE"
                        onChange={handleChange}
                      />
                    </div>
                  </InputGroup>
                </Form.Group>
              </Row>
            </Form.Row>
            <Row className="col-12 mx-0 my-5">
              <Button className="mx-auto" type="submit" disabled={isSubmitting}>Submit survey</Button>
            </Row>
          </Form>
        )}
    </Formik>
  );
}

SurveyForm.proptypes = {
  handleSubmit: PropTypes.func
}