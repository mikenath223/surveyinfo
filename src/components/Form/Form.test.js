import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { fireEvent, waitFor } from "@testing-library/react";
import { withThemeProvider } from 'utils/testUtils';
import Form from "./index";

describe('Form', function () {
  let wrapper, getTestId;

  beforeEach(() => {
    const { container, getByTestId } = withThemeProvider(<Form />);
    
    wrapper = container;
    getTestId = getByTestId;
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should show validation on blur", async () => {
    const input = getTestId("messageInput");
    
    fireEvent.blur(input);
    await waitFor(() => {
      expect(getTestId("messageError")).not.toBe(null);
      expect(getTestId("messageError")).toHaveTextContent("message is a required field");
    });
  });
})