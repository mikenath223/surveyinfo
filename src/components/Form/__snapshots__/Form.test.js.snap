// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<div>
  <form
    class="ml-lg-5 m-4"
    novalidate=""
  >
    <h2
      class="sc-bdfBwQ ggakCv"
    >
      Take the survey
    </h2>
    <div
      class="pb-4 mx-0 form-group row"
    >
      <label
        class="form-label"
        for="validationFormikMessage1"
      >
        Message
      </label>
      <div
        class="input-group"
      >
        <textarea
          class="mh-100 form-control"
          data-testid="messageInput"
          id="validationFormikMessage1"
          name="message"
          rows="7"
          spellcheck="true"
        />
        <div
          class="invalid-tooltip"
          data-testid="messageError"
        />
        <div
          class="valid-tooltip"
        >
          Looks good!
        </div>
      </div>
    </div>
    <div
      class="pb-4 mx-0 form-group row"
    >
      <label
        class="col-12 pl-0 form-label"
        for="validationFormikDate2"
      >
        Date
      </label>
      <div
        class="p-0 col-5 input-group"
      >
        <input
          class="form-control"
          id="validationFormikDate2"
          name="date"
          type="date"
          value=""
        />
        <div
          class="valid-tooltip"
        >
          Looks good!
        </div>
        <div
          class="invalid-tooltip"
        />
      </div>
    </div>
    <div
      class="pb-4 px-0 mx-0 mt-4 form-row"
    >
      <div
        class="col-12 pl-0 form-label col"
      >
        Customize your theme
      </div>
      <div
        class="col-12 mx-0 border border-1 mt-1 p-2 row"
      >
        <div
          class="form-group col-md-4"
        >
          <label
            class="font-weight-bold form-label"
          >
            Layout
          </label>
          <div
            class="input-group"
          >
            <div
              aria-labelledby="my-radio-group"
              role="group"
            >
              <div
                class="form-check"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="layout"
                  name="layout"
                  type="radio"
                  value="LIST"
                />
                <label
                  class="form-check-label"
                  for="layout"
                  title=""
                >
                  List
                </label>
              </div>
              <div
                class="form-check"
              >
                <input
                  class="form-check-input"
                  id="layout"
                  name="layout"
                  type="radio"
                  value="GRID"
                />
                <label
                  class="form-check-label"
                  for="layout"
                  title=""
                >
                  Grid
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="form-group col-md-4"
        >
          <label
            class="font-weight-bold form-label"
          >
            Color Mode
          </label>
          <div
            class="input-group"
          >
            <div
              aria-labelledby="my-radio-group"
              role="group"
            >
              <div
                class="form-check"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="color"
                  name="color"
                  type="radio"
                  value="LIGHT"
                />
                <label
                  class="form-check-label"
                  for="color"
                  title=""
                >
                  Light Mode
                </label>
              </div>
              <div
                class="form-check"
              >
                <input
                  class="form-check-input"
                  id="color"
                  name="color"
                  type="radio"
                  value="DARK"
                />
                <label
                  class="form-check-label"
                  for="color"
                  title=""
                >
                  Dark Mode
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="form-group col-md-4"
        >
          <label
            class="font-weight-bold form-label"
          >
            Text Size
          </label>
          <div
            class="input-group"
          >
            <div
              aria-labelledby="my-radio-group"
              role="group"
            >
              <div
                class="form-check"
              >
                <input
                  checked=""
                  class="form-check-input"
                  id="text"
                  name="text"
                  type="radio"
                  value="REGULAR"
                />
                <label
                  class="form-check-label"
                  for="text"
                  title=""
                >
                  Regular
                </label>
              </div>
              <div
                class="form-check"
              >
                <input
                  class="form-check-input"
                  id="text"
                  name="text"
                  type="radio"
                  value="LARGE"
                />
                <label
                  class="form-check-label"
                  for="text"
                  title=""
                >
                  Large
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-12 mx-0 my-5 row"
    >
      <button
        class="mx-auto btn btn-primary"
        type="submit"
      >
        Submit survey
      </button>
    </div>
  </form>
</div>
`;
